client/implementation/src/main/java/io/smallrye/graphql/client/typesafe/impl/HeaderBuilder.java

```diff
@@ -0,0 +1,117 @@
+package io.smallrye.graphql.client.typesafe.impl;
+
+import static io.smallrye.graphql.client.typesafe.impl.CollectionUtils.toMultivaluedMap;
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import java.util.AbstractMap.SimpleEntry;
+import java.util.Base64;
+
+import javax.ws.rs.core.MultivaluedMap;
+
+import org.eclipse.microprofile.config.Config;
+import org.eclipse.microprofile.config.ConfigProvider;
+
+import io.smallrye.graphql.client.typesafe.api.AuthorizationHeader;
+import io.smallrye.graphql.client.typesafe.api.GraphQlClientApi;
+import io.smallrye.graphql.client.typesafe.api.GraphQlClientException;
+import io.smallrye.graphql.client.typesafe.api.Header;
+import io.smallrye.graphql.client.typesafe.impl.reflection.MethodInfo;
+import io.smallrye.graphql.client.typesafe.impl.reflection.MethodResolver;
+import io.smallrye.graphql.client.typesafe.impl.reflection.TypeInfo;
+
+public class HeaderBuilder {
+    private final MethodInfo method;
+
+    public HeaderBuilder(MethodInfo method) {
+        this.method = method;
+    }
+
+    public MultivaluedMap<String, Object> build() {
+        MultivaluedMap<String, Object> headers = method.getResolvedAnnotations(Header.class)
+                .map(header -> new SimpleEntry<>(header.name(), resolveValue(header)))
+                .collect(toMultivaluedMap());
+        method.parameters()
+                .filter(parameter -> parameter.isAnnotated(Header.class))
+                .forEach(parameter -> {
+                    Header header = parameter.getAnnotations(Header.class)[0];
+                    headers.add(header.name(), parameter.getValue());
+                });
+        method.getResolvedAnnotations(AuthorizationHeader.class)
+                .findFirst()
+                .map(header -> resolveAuthHeader(method.getDeclaringType(), header))
+                .ifPresent(auth -> headers.add("Authorization", auth));
+        return headers;
+    }
+
+    private Object resolveValue(Header header) {
+        if (!header.method().isEmpty()) {
+            if (!header.constant().isEmpty())
+                throw new GraphQlClientException("Header with 'method' AND 'constant' not allowed: " + header);
+            return resolveMethodValue(header.method());
+        }
+        if (header.constant().isEmpty())
+            throw new GraphQlClientException("Header must have either 'method' XOR 'constant': " + header);
```

Comment:
Just a thought here - what do you think about using the constant as a fallback in case the method either throws an exception or returns null?

Reply:
Hmmm... I'd say that should be the responsibility of the method to handle, shouldn't it?

Reply:
either way works for me - if you allow the constant to be the fallback, that might simplify code where different interfaces could specify different fallbacks - maybe something like this:

```
@Header(name="ImportantHeader", method="getHeaderFromDB", constant="123")
Widget getWidget(String criteria...);

@Header(name="ImportantHeader", method="getHeaderFromDB", constant="456")
Widget getADifferentWidget(String criteria...);

static String getHeaderFromDB() {
    if (canAccessDB()) {
        return headerFromDB();
    }
    return null; // or throw exception
}
```

But we can do it either way.